# How to use danda
The first step you need to do is to collect analysis, study the report, then decide if you want to go with the automatic danda improvements

In [1]:
#from pprint import pprint
#import numpy as np
import pandas as pd
# notice the need to danda import to have accessor df
import danda # noqa: F401  # registers the pandas accessor
from danda.report_renderer import ReportRenderer

In [2]:
df = pd.read_csv("https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv")
df

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
...,...,...,...,...,...,...,...,...,...,...,...,...
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,NaN,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.0000,B42,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.4500,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,C148,C


In [3]:
report = df.dg.clean().dg.optimize().dg.analyze().dg.report
renderer = ReportRenderer()
print(renderer.render(report))
#print(json.dumps(report, indent=4, default=str))

Danda Report

Clean
-----

EmptySpacesPlugin
    Stripped leading/trailing whitespace from columns: Name

EmptyRowsPlugin
    Number of deleted rows: 0

EmptyColumnsPlugin
    Number of deleted columns: 0

DropDuplicates
    Number of deleted rows: 0

Execution
---------
Plugins executed : 7
Plugin names     : EmptySpacesPlugin, NormalizeMissingValuesPlugin, SparseColumnsPlugin, SparseRowsPlugin, EmptyRowsPlugin, EmptyColumnsPlugin, DropDuplicates

Optimize
--------

BooleanTypePlugin
    convert these columns to boolean Survived

CategoryTypePlugin
    Converted the following columns to category: Pclass, Sex, Age, SibSp, Parch, Embarked

Execution
---------
Plugins executed : 4
Plugin names     : BooleanTypePlugin, DateTimeTypePlugin, NumericTypePlugin, CategoryTypePlugin
Memory before    : 121,744 bytes
Memory after     : 74,678 bytes
Memory saved     : 47,066 bytes (38.7%)

Analyze
-------

ColumnSummary
    Column Summary:
    
    PassengerId
    Type: int64
    Missing: 0
    Uni

Looking at the report we can see:
* Stripped leading/trailing whitespace from columns: Name
* convert these columns to boolean Survived
* Converted the following columns to category: Pclass, Sex, Age, SibSp, Parch, Embarked
* Memory saved     : 47,066 bytes (38.7%)

Looking at the analysis:
* we have 891 rows and 12 columns
* age has 177missing values
* cabin has 687 missing values

This is good so let us do it



In [5]:
new_df = df.dg.clean().dg.optimize()


## Let us verify

In [6]:
print(df.info())
print(new_df.info())

<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    str    
 4   Sex          891 non-null    str    
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    str    
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    str    
 11  Embarked     889 non-null    str    
dtypes: float64(2), int64(5), str(5)
memory usage: 118.9 KB
None
<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   PassengerId  891 non-null    int64   
 1   Survived     891 non-null